Using inline trait bounds instead of a where clause triggers build failure

### Problem

Inline trait bound for a service - `impl<T: TestTrait> TestTraitBoundsService<T>` - fails to compile. Log attached.



### Steps

1. cargo-sails new-program test-trait-bounds
2. 
```rust
struct TestTraitBoundsService<T>(PhantomData<T>);

trait TestTrait {}

impl TestTrait for u8 {}

#[sails_rs::service]
impl<T: TestTrait> TestTraitBoundsService<T> /* ... */
```
3.
```rust
// Exposed service
    pub fn test_trait_bounds(&self) -> TestTraitBoundsService<u8> {
        TestTraitBoundsService::new()
    }
```
4. cargo test


### Possible Solution

_No response_

### Notes

_No response_

### Relevant Log Output

<details><summary>cargo test output</summary>
<p>

```
$ cargo test
    Blocking waiting for file lock on package cache
    Blocking waiting for file lock on package cache
    Blocking waiting for file lock on package cache
   Compiling test-trait-bounds-app v0.1.0 (/home/user/devel/gear-tech/sails/test-trait-bounds/app)
error[E0599]: the method `do_something` exists for struct `Box<TestTraitBoundsService<T>>`, but its trait bounds were not satisfied
  --> app/src/lib.rs:20:12
   |
20 |     pub fn do_something(&mut self) -> String {
   |            ^^^^^^^^^^^^ method cannot be called on `Box<TestTraitBoundsService<T>>` due to unsatisfied trait bounds
   |
note: trait bound `T: TestTrait` was not satisfied
  --> app/src/lib.rs:13:9
   |
13 | impl<T: TestTrait> TestTraitBoundsService<T>
   |         ^^^^^^^^^  -------------------------
   |         |
   |         unsatisfied trait bound introduced here

For more information about this error, try `rustc --explain E0599`.
error: could not compile `test-trait-bounds-app` (lib) due to 1 previous error
warning: build failed, waiting for other jobs to finish...
error: could not compile `test-trait-bounds-app` (lib) due to 1 previous error
```

</p>
</details>

--- a/rs/macros/core/src/program/mod.rs
+++ b/rs/macros/core/src/program/mod.rs
@@ -125,8 +125,8 @@ fn gen_gprogram_impl(program_impl: ItemImpl, program_args: ProgramArgs) -> Token
 
     let services_routes = services_data.iter().map(|item| &item.0);
 
-    let (program_type_path, program_type_args) = shared::impl_type(&program_impl);
-    let program_type_constraints = shared::impl_constraints(&program_impl);
+    let (program_type_path, _program_type_args) = shared::impl_type(&program_impl);
+    let (generics, program_type_constraints) = shared::impl_constraints(&program_impl);
 
     let (ctors_data, init_fn) = generate_init(
         &mut program_impl,
@@ -148,7 +148,7 @@ fn gen_gprogram_impl(program_impl: ItemImpl, program_args: ProgramArgs) -> Token
 
         #program_impl
 
-        impl #program_type_args #sails_path::meta::ProgramMeta for #program_type_path #program_type_constraints {
+        impl #generics #sails_path::meta::ProgramMeta for #program_type_path #program_type_constraints {
             fn constructors() -> #scale_info_path::MetaType {
                 #scale_info_path::MetaType::new::<meta_in_program::ConstructorsMeta>()
             }

--- a/rs/macros/core/src/service/mod.rs
+++ b/rs/macros/core/src/service/mod.rs
@@ -103,7 +103,7 @@ fn generate_gservice(args: TokenStream, service_impl: ItemImpl) -> TokenStream {
     let scale_info_path = sails_paths::scale_info_path(&sails_path);
 
     let (service_type_path, service_type_args) = shared::impl_type(&service_impl);
-    let service_type_constraints = shared::impl_constraints(&service_impl);
+    let (generics, service_type_constraints) = shared::impl_constraints(&service_impl);
 
     let service_handlers = discover_service_handlers(&service_impl);
 
@@ -319,7 +319,7 @@ fn generate_gservice(args: TokenStream, service_impl: ItemImpl) -> TokenStream {
         #exposure_drop_code
 
         #( #exposure_allow_attrs )*
-        impl #service_type_args Exposure< #exposure_args > #service_type_constraints {
+        impl #generics Exposure< #exposure_args > #service_type_constraints {
             #( #exposure_funcs )*
 
             #( #base_exposure_accessors )*
@@ -341,7 +341,7 @@ fn generate_gservice(args: TokenStream, service_impl: ItemImpl) -> TokenStream {
             #exposure_set_event_listener_code
         }
 
-        impl #service_type_args #sails_path::gstd::services::Exposure for Exposure< #exposure_args > #service_type_constraints {
+        impl #generics #sails_path::gstd::services::Exposure for Exposure< #exposure_args > #service_type_constraints {
             fn message_id(&self) -> #sails_path::MessageId {
                 self. #message_id_ident
             }
@@ -351,7 +351,7 @@ fn generate_gservice(args: TokenStream, service_impl: ItemImpl) -> TokenStream {
             }
         }
 
-        impl #service_type_args #sails_path::gstd::services::Service for #service_type_path #service_type_constraints {
+        impl #generics #sails_path::gstd::services::Service for #service_type_path #service_type_constraints {
             type Exposure = Exposure< #exposure_args >;
 
             fn expose(self, #message_id_ident : #sails_path::MessageId, #route_ident : &'static [u8]) -> Self::Exposure {
@@ -375,7 +375,7 @@ fn generate_gservice(args: TokenStream, service_impl: ItemImpl) -> TokenStream {
             }
         }
 
-        impl #service_type_args #sails_path::meta::ServiceMeta for #service_type_path #service_type_constraints {
+        impl #generics #sails_path::meta::ServiceMeta for #service_type_path #service_type_constraints {
             fn commands() -> #scale_info_path ::MetaType {
                 #scale_info_path ::MetaType::new::<meta_in_service::CommandsMeta>()
             }

--- a/rs/macros/core/src/shared.rs
+++ b/rs/macros/core/src/shared.rs
@@ -4,9 +4,10 @@ use proc_macro_error::abort;
 use quote::{quote, ToTokens};
 use std::collections::BTreeMap;
 use syn::{
-    punctuated::Punctuated, spanned::Spanned, FnArg, GenericArgument, Ident, ImplItem, ImplItemFn,
-    ItemImpl, Lifetime, Pat, Path, PathArguments, PathSegment, Receiver, ReturnType, Signature,
-    Token, Type, TypeImplTrait, TypeParamBound, TypePath, TypeReference, TypeTuple, WhereClause,
+    punctuated::Punctuated, spanned::Spanned, FnArg, GenericArgument, Generics, Ident, ImplItem,
+    ImplItemFn, ItemImpl, Lifetime, Pat, Path, PathArguments, PathSegment, Receiver, ReturnType,
+    Signature, Token, Type, TypeImplTrait, TypeParamBound, TypePath, TypeReference, TypeTuple,
+    WhereClause,
 };
 
 pub(crate) fn impl_type(item_impl: &ItemImpl) -> (TypePath, PathArguments) {
@@ -24,8 +25,10 @@ pub(crate) fn impl_type(item_impl: &ItemImpl) -> (TypePath, PathArguments) {
     (path, args)
 }
 
-pub(crate) fn impl_constraints(item_impl: &ItemImpl) -> Option<WhereClause> {
-    item_impl.generics.where_clause.clone()
+pub(crate) fn impl_constraints(item_impl: &ItemImpl) -> (Generics, Option<WhereClause>) {
+    let mut generics = item_impl.generics.clone();
+    let where_clause = generics.where_clause.take();
+    (generics, where_clause)
 }
 
 /// Represents parts of a handler function.

--- a/rs/macros/tests/gservice_tests.rs
+++ b/rs/macros/tests/gservice_tests.rs
@@ -8,6 +8,7 @@ mod gservice_with_extends_and_lifetimes;
 mod gservice_with_lifecycles_and_generics;
 mod gservice_with_lifetimes_and_events;
 mod gservice_with_reply_with_value;
+mod gservice_with_trait_bounds;
 
 #[tokio::test]
 async fn gservice_with_basics() {
@@ -308,6 +309,27 @@ async fn gservice_with_reply_with_value_with_impl_from() {
 
     let result = String::decode(&mut output).unwrap();
     assert_eq!(result, "42: correct");
+
+    assert_eq!(output.len(), 0);
+}
+
+#[tokio::test]
+async fn gservice_with_trait_bounds() {
+    use gservice_with_trait_bounds::MyServiceWithTraitBounds;
+
+    const DO_THIS: &str = "DoThis";
+
+    let (output, _value) = MyServiceWithTraitBounds::<u32>::default()
+        .expose(MessageId::from(123), &[1, 2, 3])
+        .handle(&DO_THIS.encode())
+        .await;
+    let mut output = output.as_slice();
+
+    let func_name = String::decode(&mut output).unwrap();
+    assert_eq!(func_name, DO_THIS);
+
+    let result = u32::decode(&mut output).unwrap();
+    assert_eq!(result, 42);
 
     assert_eq!(output.len(), 0);
 }

--- a/rs/macros/tests/gservice_with_trait_bounds/mod.rs
+++ b/rs/macros/tests/gservice_with_trait_bounds/mod.rs
@@ -1,0 +1,14 @@
+use core::marker::PhantomData;
+use sails_rs::gstd::service;
+
+#[derive(Default)]
+pub(super) struct MyServiceWithTraitBounds<'a, T> {
+    _a: PhantomData<&'a T>,
+}
+
+#[service]
+impl<'a, T: Into<u32>> MyServiceWithTraitBounds<'a, T> {
+    pub fn do_this(&mut self) -> u32 {
+        42
+    }
+}

